### DISCUSSION TITLE
No multiple inheritance - ways to get same outcome without it

### DISCUSSION THREAD

**Discussion by GarikFirst** [micropython/micropython#13729] [General]
I would like to do something like this:
```

from machine import ADC, Pin


class BasicSensor:
    def __init__(self) -> None:
        # some common attributes
        pass


class OnOffSensor(BasicSensor, Pin):
    def __init__(self) -> None:
        # some special attributes
        super().__init__()

    # some special functions


class AnalogSensor(BasicSensor, ADC):
    def __init__(self) -> None:
        # some special attributes
        super().__init__()

    # some special functions

```

But MicroPython don’t support multiple inheritance and call only first super(). Is there any pattern/technic that allow me to achieve same result without multiple inheritance?

**Comment by heckmon** [comment]
So we can't use subclasses in micropython??

**Reply by karfas** [reply]
You *can* use subclasses, but inheriting from **multiple** base classes is (at least) problematic.
See https://docs.micropython.org/en/v1.9.3/pyboard/genrst/core_language.html#when-inheriting-from-multiple-classes-super-only-calls-one-class.

**Reply by heckmon** [reply]
Got it thanks!

**Comment by bixb922** [comment]
Another pattern useful in this context is composition, as in https://en.m.wikipedia.org/wiki/Composition_over_inheritance

**Comment by peterhinch** [comment]
Re multiple inheritance the [docs](https://github.com/micropython/micropython/wiki/Differences) state:
> MicroPython does not implement complete CPython object data model, but only a subset of it. Advanced usages of multiple inheritance, __new__ method may not work. Method resolution order is different (#525).

Translating, MI does work in MicroPython provided you don't try to be too clever. Note the link given by @karfas.

There is usually a better way than using MI. In nearly ten years of MP coding I used it once.

**Comment by GarikFirst** [comment]
I think I can close this now, composition will do the job
